File: android/app/src/test/java/org/switchboard/android/NetworksTest.kt
package org.switchboard.android

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * The list offered to somebody with no networks.
 *
 * This phone reads the desktop's `src/shared/networks.json`, copied into its
 * assets at build time. The copy is what this checks: that the file the build
 * hands the app is the one the desktop ships, and that it parses into the
 * shape the picker expects.
 *
 * A broken entry here is a first-run experience that dead-ends, on the one
 * screen where a user has no way to tell our mistake from their own.
 */
class NetworksTest {

    private val json = Json { ignoreUnknownKeys = true }

    private val original = File(
        System.getProperty("switchboard.fixtures")!!
    ).resolveSibling("../src/shared/networks.json").normalize()

    private val copied = File("src/main/assets/networks.json")

    private fun parse(file: File) =
        json.parseToJsonElement(file.readText()).let { it as JsonObject }

    @Test
    fun `the asset is the desktop's file, not a second copy of it`() {
        assertTrue("the build should have copied it: $copied", copied.exists())
        assertEquals(original.readText(), copied.readText())
    }

    @Test
    fun `parses into what the picker renders`() {
        val networks = parse(copied)["networks"]!!.jsonArray
        assertTrue("a useful number of them", networks.size >= 8)

        for (entry in networks) {
            val network = entry as JsonObject
            val id = network["id"]!!.jsonPrimitive.content

            assertTrue("$id has a name", network["name"]!!.jsonPrimitive.content.isNotBlank())
            assertTrue("$id says what it is for", network["description"]!!.jsonPrimitive.content.length > 20)
            assertTrue("$id says where it is", network["region"]!!.jsonPrimitive.content.isNotBlank())
            assertTrue("$id has a host", network["host"]!!.jsonPrimitive.content.contains('.'))
            assertTrue("$id is over TLS", network["tls"]!!.jsonPrimitive.content == "true")

            val port = network["port"]!!.jsonPrimitive.content.toInt()
            assertTrue("$id has a usable port", port in 1..65535)
        }
    }

    @Test
    fun `no two of them are the same network`() {
        val networks = parse(copied)["networks"]!!.jsonArray
        val ids = networks.map { (it as JsonObject)["id"]!!.jsonPrimitive.content }
        val hosts = networks.map { (it as JsonObject)["host"]!!.jsonPrimitive.content }

        assertEquals(ids.size, ids.toSet().size)
        assertEquals(hosts.size, hosts.toSet().size)
    }
}

File: tests/main/networks.test.ts
import { describe, it, expect } from 'vitest'
import { KNOWN_NETWORKS, NETWORKS_CHECKED_AT, suggestedNetworks } from '@shared/networks'

/**
 * The list offered to somebody with no networks.
 *
 * It is data rather than logic, so what is worth testing is that it stays
 * usable: a broken entry here is a first-run experience that dead-ends, and
 * this is the one screen where a user has no way to tell our mistake from
 * their own.
 *
 * The Android app reads the same file — `src/shared/networks.json`, copied into
 * its assets at build time — so there is nothing here for the two to disagree
 * about.
 */
describe('the networks we offer to start with', () => {
  it('has a useful number of them', () => {
    expect(KNOWN_NETWORKS.length).toBeGreaterThanOrEqual(8)
  })

  it('says when it was last checked', () => {
    expect(NETWORKS_CHECKED_AT).toMatch(/^\d{4}-\d{2}-\d{2}$/)
  })

  it('gives every one a name, a description and a place', () => {
    for (const network of KNOWN_NETWORKS) {
      expect(network.name, network.id).toBeTruthy()
      expect(network.description.length, network.id).toBeGreaterThan(20)
      expect(network.region, network.id).toBeTruthy()
    }
  })

  it('has no two the same', () => {
    const ids = KNOWN_NETWORKS.map((n) => n.id)
    expect(new Set(ids).size).toBe(ids.length)

    const hosts = KNOWN_NETWORKS.map((n) => n.host)
    expect(new Set(hosts).size).toBe(hosts.length)
  })

  it('points every one at a hostname and a port', () => {
    for (const network of KNOWN_NETWORKS) {
      expect(network.host, network.id).toMatch(/^[a-z0-9.-]+\.[a-z]{2,}$/)
      expect(network.port, network.id).toBeGreaterThan(0)
      expect(network.port, network.id).toBeLessThan(65536)
    }
  })

  /**
   * Nobody should be sent to a plain-text network by a list we wrote. Where a
   * network's own round-robin cannot be verified, the entry points at a server
   * that can — see IRCnet and EFnet.
   */
  it('sends nobody anywhere unencrypted', () => {
    for (const network of KNOWN_NETWORKS) {
      expect(network.tls, network.id).toBe(true)
    }
  })

  it('records what answered when each was checked', () => {
    for (const network of KNOWN_NETWORKS) {
      expect(network.checked, network.id).toBeTruthy()
    }
  })

  it('suggests channels that are channels', () => {
    for (const network of KNOWN_NETWORKS) {
      for (const channel of network.channels ?? []) {
        expect(channel, network.id).toMatch(/^#/)
      }
    }
  })

  // ── searching it ─────────────────────────────────────────────────

  it('offers everything when nothing has been typed', () => {
    expect(suggestedNetworks('')).toHaveLength(KNOWN_NETWORKS.length)
    expect(suggestedNetworks('   ')).toHaveLength(KNOWN_NETWORKS.length)
  })

  it('finds one by name', () => {
    expect(suggestedNetworks('libera').map((n) => n.id)).toEqual(['libera'])
  })

  it('finds them by what they are for', () => {
    expect(suggestedNetworks('anime').map((n) => n.id)).toContain('rizon')
    expect(suggestedNetworks('debian').map((n) => n.id)).toContain('oftc')
  })

  it('finds them by where they are', () => {
    const european = suggestedNetworks('europe').map((n) => n.id)
    expect(european).toContain('ircnet')
    expect(european).toContain('quakenet')
  })

  it('finds one by its address', () => {
    expect(suggestedNetworks('oftc.net').map((n) => n.id)).toEqual(['oftc'])
  })

  it('returns nothing for something that is not there', () => {
    expect(suggestedNetworks('zzzznotanetwork')).toEqual([])
  })
})
